storage-proofs/porep/src/nse/vanilla/labels.rs

```diff
@@ -493,6 +495,66 @@ fn tree_from_slice<Tree: MerkleTreeTrait>(
     Ok(tree)
 }
 
+fn to_gpu_config(conf: &Config) -> gpu::Config {
+    gpu::Config {
+        num_nodes_window: conf.num_nodes_window,
+        num_butterfly_layers: conf.num_butterfly_layers,
+        num_expander_layers: conf.num_expander_layers,
+        degree_expander: conf.degree_expander,
+        degree_butterfly: conf.degree_butterfly,
+        k: conf.k,
+    }
+}
+
+type GPUHasherDomain = storage_proofs_core::hasher::PoseidonDomain;
+type GPUHasher = storage_proofs_core::hasher::PoseidonHasher;
+type GPUTree = storage_proofs_core::merkle::OctLCMerkleTree<GPUHasher>;
+pub fn encode_with_oct_lc_poseidon_trees_gpu(
+    conf: &Config,
+    mut store_configs: Vec<StoreConfig>,
+    window_index: u32,
+    replica_id: &GPUHasherDomain,
+    data: &mut [u8],
+) -> gpu::NSEResult<(Vec<MerkleTree<GPUTree>>, Option<LCMerkleTree<GPUTree>>)> {
+    use storage_proofs_core::fr32::fr_into_bytes;
+    let gpu_conf = to_gpu_config(conf);
+    let mut context = gpu::GPU::new(gpu_conf)?;
+    let sealer = gpu::Sealer::new(
+        gpu_conf,
+        unsafe { std::mem::transmute::<_, gpu::ReplicaId>(*replica_id) },
+        window_index as usize,
+        gpu::Layer::from(&data.to_vec()),
+        &mut context,
+        true,
+        0,
+    )?;
+    let layers = sealer
+        .map(|v| v)
+        .collect::<gpu::NSEResult<Vec<gpu::LayerOutput>>>()?;
+    data.copy_from_slice(Vec::<u8>::from(&layers.last().unwrap().base).as_slice());
+    let mut trees = Vec::new();
+    for lo in layers.iter() {
+        let store_config = store_configs.remove(0);
+        let tree_data: Vec<u8> = lo
+            .base
+            .0
+            .iter()
+            .chain(lo.tree.iter())
+            .flat_map(|node| fr_into_bytes(&node.0))
+            .collect();
+        trees.push(
+            MerkleTree::<GPUTree>::from_tree_slice_with_config(
```

Comment:
@dignifiedquire This is not working properly atm, how can I build a `MerkleTree::<GPUTree>` given a `Vec<Fr>` returned from Neptune? (And what should I set `discard_rows` in case I want to do this?